File: src/mapper.cpp
#include <algorithm>

#include "mapper.h"


Mapper0::Mapper0()
{
}

Mapper0::Mapper0(std::vector<uint8_t>&& prg_rom, std::vector<uint8_t>&& chr_rom) :
    prg_rom(std::move(prg_rom)),
    chr_rom(std::move(chr_rom))
{
}

Mapper0& Mapper0::operator=(Mapper0&& mapper)
{
    if (this == &mapper)
    {
        return *this;
    }
    std::swap(prg_rom, mapper.prg_rom);
    std::swap(chr_rom, mapper.chr_rom);
    return *this;
}

uint8_t Mapper0::read(uint16_t address) const
{
    if (address >= 0x8000 && address <= 0xFFFF)
    {
        return prg_rom[address & 0x3FFF];
    }
    if (address >= 0x0000 && address <= 0x1FFF)
    {
        return chr_rom[address];
    }
}

void Mapper0::write(uint16_t address, uint8_t data)
{
}

File: src/mapper.rs
#[derive(Debug)]
pub struct Mapper0 {
    prg_rom: Vec<u8>,
    chr_rom: Vec<u8>,
}

impl Mapper0 {
    pub fn new(prg_rom: Vec<u8>, chr_rom: Vec<u8>) -> Self {
        Self { prg_rom, chr_rom }
    }

    pub fn read(&self, address: u16) -> u8 {
        match address {
            // this actual for 1 bank roms (mirrored) but not for 2 banks
            0x8000..=0xFFFF => self.prg_rom[(address & 0x3FFF) as usize],
            0x0000..=0x1FFF => self.chr_rom[address as usize],
            _ => panic!("invalid address {:#X}", address),
        }
    }
}
